src/os/bluestore/BlueFS.cc

```diff
@@ -3038,54 +3038,74 @@ void BlueFS::_consume_dirty(uint64_t seq)
   }
 }
 
-// Extends log if its free space is smaller then bluefs_min_log_runway.
-// Returns space available *BEFORE* adding new space. Signed for additional <0 detection.
-int64_t BlueFS::_maybe_extend_log()
-{
+int64_t BlueFS::_maybe_extend_log(uint64_t amount, bool ignore_compaction) {
+  uint64_t runway = log.writer->file->fnode.get_allocated() - log.writer->get_effective_write_pos();
+  // increasing the size of the log involves adding a OP_FILE_UPDATE_INC which its size will 
+  // increase with respect the number of extents. bluefs_min_log_runway should cover the max size 
+  // a log can get.
+  if (runway < cct->_conf->bluefs_min_log_runway) {
+    _extend_log(log.t.expected_size(), ignore_compaction);
+  }
+  // inject new allocation in case log is too big
+  if (log.t.expected_size() > runway) {
+    _extend_log(log.t.expected_size(), ignore_compaction);
+  }
+  runway = log.writer->file->fnode.get_allocated() - log.writer->get_effective_write_pos();
+  return runway;
+}
+
+int64_t BlueFS::_extend_log(uint64_t amount, bool ignore_compaction) {
   ceph_assert(ceph_mutex_is_locked(log.lock));
-  // allocate some more space (before we run out)?
-  // BTW: this triggers `flush()` in the `page_aligned_appender` of `log.writer`.
-  int64_t runway = log.writer->file->fnode.get_allocated() -
-    log.writer->get_effective_write_pos();
-  if (runway < (int64_t)cct->_conf->bluefs_min_log_runway) {
-    dout(10) << __func__ << " allocating more log runway (0x"
-	     << std::hex << runway << std::dec  << " remaining)" << dendl;
-    /*
-     * Usually, when we are low on space in log, we just allocate new extent,
-     * put update op(log) to log and we are fine.
-     * Problem - it interferes with log compaction:
-     * New log produced in compaction will include - as last op - jump into some offset (anchor) of current log.
-     * It is assumed that log region (anchor - end) will contain all changes made by bluefs since
-     * full state capture into new log.
-     * Putting log update into (anchor - end) region is illegal, because any update there must be compatible with
-     * both logs, but old log is different then new log.
-     *
-     * Possible solutions:
-     * - stall extending log until we finish compacting and switch log (CURRENT)
-     * - re-run compaction with more runway for old log
-     * - add OP_FILE_ADDEXT that adds extent; will be compatible with both logs
-     */
-    if (log_forbidden_to_expand.load() == true) {
-      return -EWOULDBLOCK;
+  if (!ignore_compaction) {
+    std::unique_lock<ceph::mutex> ll(log.lock, std::adopt_lock);
+    while (log_forbidden_to_expand.load() == true) {
+      log_cond.wait(ll);
     }
-    vselector->sub_usage(log.writer->file->vselector_hint, log.writer->file->fnode);
+    ll.release();
```

Comment:
What is the best way to deal with the lock in this case? This while loop unlocks the lock in case of `log_forbidden_to_expand` and locks it back again when the value changes. The release is there because even though we used the log.lock mutex in `log_cond` we don't want to unlock when `ll` leaves the block scope.